### PR TITLE
MC-35574: show source stock instead of default stock

### PR DESCRIPTION
**Pull request by tsviklinskyi**
### Related Pull Requests
https://github.com/magento/magento2ee/pull/2516
https://github.com/magento/magento2-infrastructure/pull/1263

### Description (*)

PREREQUISITES: clean Magento 2.4-develop + MSI

ISSUE:  
default stock of products is visible instead of source2 related to website2 on product sorting with stock in Visual merchandiser 
"Move low stock to top" sort is based on default source instead of source2 related to website2.

STEPS TO REPLICATE:
Login to Magento Admin.
Set up multi website and assign different sources to different websites
-Default Website+Default Store+Default Storeview=Default Source+Default Stock
-Website2+Store2+Storeview2=Souce2+Stock2
Create category and simple product. Assign this product to this category and both websites and both stocks as follows:
-Default Source+Default Stock=QTY=0
-Souce2+Stock2=QTY=100
Go to Catalog>Category
Select Storeview2 and click on category
Click "products in category", Set Tile view instead of Grid view option and select sort option as "Move low stock to top"

ACTUAL RESULTS:
the stock source is not visible just the default one:
-We can see default stock=0 only stock2=100 is not visible
By selecting sort the products with option "Move low stock to top" I see the products sort is  based on default stock instead of source stock2 assigned to website2

EXPECTED RESULTS:
Both (like on Catalog>>Products) or at least the related stock source2 should be visible
Product sort  should be based on source stock2 assigned to website2.

### Fixed Issues (if relevant)

1. Fixes magento/inventory#<issue_number>: Issue title

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by sidolov** [comment]
@magento run all tests

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Comment by smoskaluk** [comment]
hi @tsviklinskyi , I tested PR and still sorting is not correct
image
image

**Comment by tsviklinskyi** [comment]
Hi @smoskaluk , sorting is now fixed.

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Comment by sidolov** [comment]
@magento run all tests

File: InventoryVisualMerchandiser/composer.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    "require": {
+        "php": "~7.3.0||~7.4.0",
+        "magento/framework": "*",
+        "magento/module-catalog": "*",
+        "magento/module-inventory-catalog-api": "*",
+        "magento/module-inventory-indexer": "*",
+        "magento/module-inventory-sales-api": "*",
+        "magento/module-store": "*"
+    },
+    "require-dev": {
```
Comment (sidolov) [inline InventoryVisualMerchandiser/composer.json:14]: Why mentioned dependencies in the required-dev section? We don't need them on the production env?

File: InventoryVisualMerchandiser/Test/Integration/Model/Category/ProductsTest.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+use Magento\TestFramework\Helper\Bootstrap;
+use Magento\VisualMerchandiser\Model\Category\Products;
+use PHPUnit\Framework\TestCase;
+
+class ProductsTest extends TestCase
+{
+    /**
+     * Test that products in category have correct quantity by source
+     *
+     * @magentoDataFixture ../../../../app/code/Magento/InventorySalesApi/Test/_files/websites_with_stores.php
```
Comment (sidolov) [inline InventoryVisualMerchandiser/Test/Integration/Model/Category/ProductsTest.php]: Please, use alias for the fixture. Example: `@magentoDataFixture Magento_InventoryApi::Test/_files/stocks.php`

File: InventoryVisualMerchandiser/Plugin/Model/Resolver/QuantityAndStockPlugin.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+namespace Magento\InventoryVisualMerchandiser\Plugin\Model\Resolver;
+
+use Magento\Catalog\Model\ResourceModel\Product\Collection;
+use Magento\Framework\Exception\LocalizedException;
+use Magento\InventoryCatalogApi\Api\DefaultStockProviderInterface;
+use Magento\InventoryIndexer\Model\StockIndexTableNameResolverInterface;
+use Magento\InventorySalesApi\Api\Data\SalesChannelInterface;
+use Magento\InventorySalesApi\Api\StockResolverInterface;
+use Magento\Store\Model\StoreManagerInterface;
+use Magento\VisualMerchandiser\Model\Resolver\QuantityAndStock;
```
Comment (sidolov) [inline InventoryVisualMerchandiser/Plugin/Model/Resolver/QuantityAndStockPlugin.php:19]: It's not acceptable to have Inventory related logic in the magento core. Please, remove `Magento\VisualMerchandiser\Model\Resolver\QuantityAndStock` and move logic to the inventory repo and handle additional cases mentioned in this PR.

**Review by sidolov — CHANGES_REQUESTED** [review]

File: InventoryVisualMerchandiser/composer.json (hunk 1 above)
Comment (tsviklinskyi) [reply InventoryVisualMerchandiser/composer.json:14]: The require-dev section specifies those dependencies that are used only in tests, according to the rules for creating a composer.json file.

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Review by sidolov — DISMISSED** [review]

**Comment by tsviklinskyi** [comment]
@magento run all tests

**Review by sidolov — DISMISSED** [review]

**Comment by rganin** [comment]
Excluded by @zakdma request.

**Comment by sidolov** [comment]
@magento run all tests

**Comment by sidolov** [comment]
- [x] MHI failures are not relevant since the global state used in test fixtures.

**Review by sidolov — APPROVED** [review]

**Pull request merged**